File: LeetCode/1462. Course Schedule IV/solution.py
class Solution:
    def checkIfPrerequisite(self, n: int, prerequisites: List[List[int]], queries: List[List[int]]) -> List[bool]:
        pre = [[False]*n for _ in range(n)]
        g = defaultdict(list)
        degree = Counter()
        for a, b in prerequisites:
            g[a].append(b)
            degree[b] += 1
        q = deque()
        for i in range(n):
            if degree[i] == 0:
                q.append(i)
        while q:
            cur = q.popleft()
            for nxt in g[cur]:
                for i in range(n):
                    if pre[i][cur]:
                        pre[i][nxt] = True
                pre[cur][nxt] = True
                degree[nxt] -= 1
                if degree[nxt] == 0:
                    q.append(nxt)
        res = []
        for a, b in queries:
            res.append(pre[a][b])
        return res


class Solution:
    def checkIfPrerequisite(self, n, prerequisites, queries):
        connected = [[False] * n for i in range(n)]

        for i, j in prerequisites:
            connected[i][j] = True

        for k in range(n):
            for i in range(n):
                for j in range(n):
                    connected[i][j] = connected[i][j] or (connected[i][k] and connected[k][j])

        return [connected[i][j] for i, j in queries]

File: LeetCode/1462. Course Schedule IV/solution.cpp
const int maxn = 100;
bitset<maxn> f[maxn];

class Solution {
public:
    vector<bool> checkIfPrerequisite(int n, vector<vector<int>>& prerequisites, vector<vector<int>>& queries) {
        for (int i = 0; i < n; i++)
            f[i] = 0;
        for (auto p : prerequisites)
            f[p[1]][p[0]] = 1;
        for (int k = 0; k < n; k++){
            for (int i = 0; i < n; i++){
                if (f[i][k])
                    f[i] = f[i] | f[k];
            }
        }
        int m = queries.size();
        vector<bool> ans(m,false);
        for (int i = 0; i < m; i++)
            if (f[queries[i][1]][queries[i][0]])
                ans[i] = true;
        return ans;
    }
};